File: backend/app/api/v1/shift_reschedules.py
from datetime import date, time
from typing import Annotated, Literal
from uuid import UUID

from fastapi import APIRouter, Depends, Query
from pydantic import BaseModel, Field
from sqlalchemy import select
from sqlalchemy.orm import Session, selectinload

from app.api.deps import get_current_admin
from app.api.v1.admin_utils import apply_pagination, count_for, pagination_meta
from app.api.v1.team_auth import apply_employee_scope, ensure_employee_access
from app.core.exceptions import ApiError
from app.core.responses import success_response
from app.database.session import get_db
from app.models import AdminUser, Employee, ShiftRescheduleRequest
from app.services.audit import record_audit_log
from app.services.employee_archive import current_employee_clause
from app.services.shift_reschedules import (
    base_day_schedule,
    create_emergency_reschedule,
    expire_company_pending,
    require_shift_reschedule_reviewer,
    review_request,
    serialize_base_day,
    serialize_shift_reschedule,
)

router = APIRouter(prefix="/shift-reschedules", tags=["shift-reschedules"])


class ShiftRescheduleReview(BaseModel):
    status: Literal["approved", "rejected"]
    review_reason: str | None = Field(default=None, max_length=1000)


class EmergencyShiftRescheduleCreate(BaseModel):
    employee_id: UUID
    work_date: date
    requested_start: time
    requested_end: time
    reason: str = Field(min_length=3, max_length=1000)


def _audit_details(row: ShiftRescheduleRequest) -> dict:
    return {
        "work_date": row.work_date.isoformat(),
        "requested_start": row.requested_start.strftime("%H:%M"),
        "requested_end": row.requested_end.strftime("%H:%M"),
        "original_start": row.original_start.strftime("%H:%M"),
        "original_end": row.original_end.strftime("%H:%M"),
        "source": row.source,
        "review_reason": row.review_reason,
        "schedule_override_id": str(row.schedule_override_id)
        if row.schedule_override_id
        else None,
    }


@router.get("")
def list_shift_reschedules(
    current_admin: Annotated[AdminUser, Depends(get_current_admin)],
    db: Annotated[Session, Depends(get_db)],
    status: str | None = None,
    employee_id: UUID | None = None,
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=50, ge=1, le=200),
):
    require_shift_reschedule_reviewer(current_admin)
    if expire_company_pending(db, current_admin.company_id):
        db.commit()
    statement = (
        select(ShiftRescheduleRequest)
        .join(Employee, Employee.id == ShiftRescheduleRequest.employee_id)
        .where(ShiftRescheduleRequest.company_id == current_admin.company_id)
        .order_by(ShiftRescheduleRequest.created_at.desc())
    )
    statement = apply_employee_scope(
        statement, db, current_admin, ShiftRescheduleRequest.employee_id
    )
    if employee_id is not None:
        ensure_employee_access(db, current_admin, employee_id)
        statement = statement.where(ShiftRescheduleRequest.employee_id == employee_id)
    else:
        # Same convention as the other request queues: archived employees are
        # hidden by default and still reachable by filtering on the employee.
        statement = statement.where(current_employee_clause())
    if status:
        statement = statement.where(ShiftRescheduleRequest.status == status)
    total = count_for(db, statement)
    rows = db.scalars(
        apply_pagination(
            statement.options(
                selectinload(ShiftRescheduleRequest.employee),
                selectinload(ShiftRescheduleRequest.reviewed_by),
                selectinload(ShiftRescheduleRequest.created_by),
            ),
            page,
            page_size,
        )
    ).all()
    return success_response(
        data=[serialize_shift_reschedule(row) for row in rows],
        meta=pagination_meta(total, page, page_size),
    )


@router.get("/day")
def shift_reschedule_day(
    current_admin: Annotated[AdminUser, Depends(get_current_admin)],
    db: Annotated[Session, Depends(get_db)],
    employee_id: UUID,
    work_date: date = Query(alias="date"),
):
    """The employee's normal shift for a date (used by the emergency form)."""
    require_shift_reschedule_reviewer(current_admin)
    employee = ensure_employee_access(db, current_admin, employee_id)
    base = base_day_schedule(db, employee, work_date)
    db.commit()
    return success_response(data=serialize_base_day(work_date, base))


@router.post("/emergency")
def create_emergency(
    payload: EmergencyShiftRescheduleCreate,
    current_admin: Annotated[AdminUser, Depends(get_current_admin)],
    db: Annotated[Session, Depends(get_db)],
):
    require_shift_reschedule_reviewer(current_admin)
    employee = ensure_employee_access(db, current_admin, payload.employee_id)
    row = create_emergency_reschedule(
        db,
        admin=current_admin,
        employee=employee,
        work_date=payload.work_date,
        requested_start=payload.requested_start,
        requested_end=payload.requested_end,
        reason=payload.reason,
    )
    record_audit_log(
        db,
        current_admin,
        "created_emergency",
        "shift_reschedule_request",
        entity_id=row.id,
        entity_name=employee.name,
        details=_audit_details(row),
    )
    db.commit()
    db.refresh(row)
    return success_response(data=serialize_shift_reschedule(row))


@router.patch("/{request_id}")
def review_shift_reschedule(
    request_id: UUID,
    payload: ShiftRescheduleReview,
    current_admin: Annotated[AdminUser, Depends(get_current_admin)],
    db: Annotated[Session, Depends(get_db)],
):
    require_shift_reschedule_reviewer(current_admin)
    row = db.scalar(
        select(ShiftRescheduleRequest)
        .where(
            ShiftRescheduleRequest.id == request_id,
            ShiftRescheduleRequest.company_id == current_admin.company_id,
        )
        .with_for_update()
    )
    if row is None:
        raise ApiError("SHIFT_RESCHEDULE_NOT_FOUND", "Shift reschedule was not found.", 404)
    employee = ensure_employee_access(db, current_admin, row.employee_id)
    review_request(
        db,
        admin=current_admin,
        row=row,
        status=payload.status,
        review_reason=payload.review_reason,
    )
    record_audit_log(
        db,
        current_admin,
        payload.status,
        "shift_reschedule_request",
        entity_id=row.id,
        entity_name=employee.name,
        details=_audit_details(row),
    )
    db.commit()
    db.refresh(row)
    return success_response(data=serialize_shift_reschedule(row))

File: frontend/admin-dashboard/src/api/shiftReschedules.ts
import { apiFetch, withQuery } from "./client";

export type ShiftRescheduleStatus = "pending" | "approved" | "rejected" | "cancelled" | "expired";

export type ShiftReschedule = {
  id: string;
  employeeId: string;
  employeeName: string;
  employeeCode?: string | null;
  workDate: string;
  requestedStart: string;
  requestedEnd: string;
  originalStart: string;
  originalEnd: string;
  durationMinutes: number;
  reason: string;
  status: ShiftRescheduleStatus;
  source: "employee" | "admin";
  createdByName?: string | null;
  reviewedByName?: string | null;
  reviewedAt?: string | null;
  reviewReason?: string | null;
  createdAt?: string | null;
};

export type ShiftRescheduleDay = {
  date: string;
  scheduledDay: boolean;
  shiftStart: string | null;
  shiftEnd: string | null;
  shiftMinutes: number;
};

type BackendShiftReschedule = {
  id: string;
  employee_id: string;
  employee_name: string;
  employee_code?: string | null;
  work_date: string;
  requested_start: string;
  requested_end: string;
  original_start: string;
  original_end: string;
  duration_minutes: number;
  reason: string;
  status: ShiftRescheduleStatus;
  source: "employee" | "admin";
  created_by_name?: string | null;
  reviewed_by_name?: string | null;
  reviewed_at?: string | null;
  review_reason?: string | null;
  created_at?: string | null;
};

type BackendShiftRescheduleDay = {
  date: string;
  scheduled_day: boolean;
  shift_start: string | null;
  shift_end: string | null;
  shift_minutes: number;
};

function mapShiftReschedule(row: BackendShiftReschedule): ShiftReschedule {
  return {
    id: row.id,
    employeeId: row.employee_id,
    employeeName: row.employee_name,
    employeeCode: row.employee_code ?? undefined,
    workDate: row.work_date,
    requestedStart: row.requested_start,
    requestedEnd: row.requested_end,
    originalStart: row.original_start,
    originalEnd: row.original_end,
    durationMinutes: row.duration_minutes,
    reason: row.reason,
    status: row.status,
    source: row.source,
    createdByName: row.created_by_name ?? undefined,
    reviewedByName: row.reviewed_by_name ?? undefined,
    reviewedAt: row.reviewed_at ?? undefined,
    reviewReason: row.review_reason ?? undefined,
    createdAt: row.created_at ?? undefined,
  };
}

export async function listShiftReschedules(
  status: ShiftRescheduleStatus | "all",
  signal?: AbortSignal,
): Promise<ShiftReschedule[]> {
  const rows = await apiFetch<BackendShiftReschedule[]>(
    withQuery("/shift-reschedules", {
      page_size: 200,
      status: status !== "all" ? status : undefined,
    }),
    { signal },
  );
  return rows.map(mapShiftReschedule);
}

export async function getShiftRescheduleDay(
  employeeId: string,
  workDate: string,
  signal?: AbortSignal,
): Promise<ShiftRescheduleDay> {
  const row = await apiFetch<BackendShiftRescheduleDay>(
    withQuery("/shift-reschedules/day", { employee_id: employeeId, date: workDate }),
    { signal },
  );
  return {
    date: row.date,
    scheduledDay: row.scheduled_day,
    shiftStart: row.shift_start,
    shiftEnd: row.shift_end,
    shiftMinutes: row.shift_minutes,
  };
}

export async function reviewShiftReschedule(
  id: string,
  input: { status: "approved" | "rejected"; reviewReason?: string },
): Promise<ShiftReschedule> {
  const row = await apiFetch<BackendShiftReschedule>(`/shift-reschedules/${id}`, {
    method: "PATCH",
    body: JSON.stringify({ status: input.status, review_reason: input.reviewReason }),
  });
  return mapShiftReschedule(row);
}

export async function createEmergencyShiftReschedule(input: {
  employeeId: string;
  workDate: string;
  requestedStart: string;
  requestedEnd: string;
  reason: string;
}): Promise<ShiftReschedule> {
  const row = await apiFetch<BackendShiftReschedule>("/shift-reschedules/emergency", {
    method: "POST",
    body: JSON.stringify({
      employee_id: input.employeeId,
      work_date: input.workDate,
      requested_start: input.requestedStart,
      requested_end: input.requestedEnd,
      reason: input.reason,
    }),
  });
  return mapShiftReschedule(row);
}

/** Same-length end time, or null if it would reach/cross midnight. */
export function sameLengthEnd(start: string, shiftMinutes: number): string | null {
  const match = /^(\d{1,2}):(\d{2})/.exec(start);
  if (!match || shiftMinutes <= 0) return null;
  const end = Number(match[1]) * 60 + Number(match[2]) + shiftMinutes;
  if (end >= 24 * 60) return null;
  return `${String(Math.floor(end / 60)).padStart(2, "0")}:${String(end % 60).padStart(2, "0")}`;
}
